lib/internal/event_target.js

```diff
@@ -286,6 +290,17 @@ class EventTarget {
     }
     type = String(type);
 
+    if (signal) {
+      if (signal.aborted) {
+        return false;
+      }
+      // TODO(benjamingr) make this weak somehow? ideally the signal would
+      // not prevent the event target from GC.
+      signal.addEventListener('abort', () => {
```

Comment:
FWIW: we should really add support for "weak" event listeners that is:

```js
signal[kAddWeakListener]('abort', () => {
  this.removeEventListener(type, listener, options);
});
```

That would _not_ retain the event target only because the signal exists (since if no one cares about the event retaining it for `abort` is mostly meaningless). I think a good path forward is:

 - Make our listener list accept `WeakRef`s to listeners
 - Put said listener behind a WeakRef
 - Make our dispatch unwrap WeakRefs when it finds them.

@addaleax @jasnell if you think that's reasonable I'm happy to do this (either here or in a new PR)

Reply:
I think that’s reasonable here, yes… it’s way more common to think that you need a weak listener than it is to actually need one, so I wouldn’t expose this as public API, but for internal usage this sounds good :)

Reply:
My only concern there would be the potential additional performance cost. If the dispatch only incurs that cost for weak handlers then I'm definitely +1

Reply:
@jasnell great, I'll do that at a separate pull request then.

@addaleax there are discussions for other use cases for this e.g. creating linked signals https://github.com/whatwg/dom/issues/920 - there is a good explanation here: https://github.com/nodejs/node/issues/35990#issuecomment-723218855

I think we're running into the same issue with AbortSignal multiple times: If the resource will no longer be aborted and we're "done" with it we need to make sure that retaining the signal does not retain the resource. With some resources (like http.request, streams in general and timers) this is a relatively easy problem since we know when abort will be meaningless but with `addEventListener` and other resources this is less clear cut (hence the advantage of a weak handler to not have to manage this).